File: LeetCode/p860.cpp
class Solution {
public:
    bool lemonadeChange(const vector<int> &bills) {
        int note[3] = {0};
        for (int i: bills) {
            if (!f(i - 5, note))
                return false;
            ++note[box(i)];
        }
        return true;
    }

private:
    static inline bool f(int change, int *note) {
        if (10 < change && note[1] > 0) {
            --note[1];
            change -= 10;
        }
        int k = change / 5;
        if (note[0] < k) {
            return false;
        }
        note[0] -= k;
        return true;
    }

    static inline int box(int i) {
        switch (i) {
            case 20:
                return 2;
            case 10:
                return 1;
            case 5:
                return 0;
            default:
                return -1;
        }
    }
};

File: LeetCode/p860.rs
impl Solution {
  pub fn lemonade_change(bills: Vec<i32>) -> bool {
    let mut change = [0; 3];
    for bill in bills.into_iter() {
      if !Self::to_change(&mut change, bill) {
        return false;
      }
    }
    true
  }

  fn to_change(change: &mut [i32], mut i: i32) -> bool {
    change[i as usize / 10] += 1;
    i -= 5;
    for j in (0..=1).rev() {
      let mut m = std::cmp::min(change[j], i / &[5, 10, 20][j]);
      i -= &[5, 10, 20][j] * m;
      change[j] -= m;
    }
    i == 0
  }
}

